File: app/src/main/java/realinterviews/CountStringInstances.java
package realinterviews;

public class CountStringInstances {
    public int countInstances(String pattern, String findIn) {

        int count = 0;
        int findInEndIndex = findIn.length() - 1;
        int patternEndIndex = pattern.length() - 1;

        for (int i = 0; i < findIn.length(); i++) {
            int thisI = i;
            for (int j = 0; j < pattern.length(); j++) {
                if (findIn.charAt(thisI) != pattern.charAt(j)) {
                    break;
                } else if (findIn.charAt(thisI) == pattern.charAt(j) && j < patternEndIndex) {
                    thisI++;
                    if (thisI <= findInEndIndex) {
                        continue;
                    } else {
                        break;
                    }
                    
                } else { // findIn.charAt(thisI) == pattern.charAt(j) && j == patternEndIndex
                    count += 1;
                }
            }
            
        }

        return count;
    }
    
}

File: app/src/main/kotlin/interviewpractice/CountStringInstances.kt
package interviewpractice;

public class CountStringInstances {
    fun countStringInstances(pattern: String, findIn: String): Int {
        var count = 0
        val findInEndIndex = findIn.length - 1
        val patternEndIndex = pattern.length - 1

        for (i in findIn.indices) {
            var thisI = i
            for (j in pattern.indices) {
                when {
                    findIn.get(thisI) != pattern.get(j) -> break
                    findIn.get(thisI) == pattern.get(j) && j < patternEndIndex -> {
                        thisI += 1
                        when (thisI <= findInEndIndex) {
                            true -> continue
                            false -> break
                        }
                    }
                    else -> count += 1
                }
            }
        }

        return count;
    }
}